Mathlib/Algebra/BigOperators/Expect.lean

```diff
@@ -0,0 +1,397 @@
+  ranges over the finite set `s` (either a `Finset` or a `Set` with a `Fintype` instance).
+* `𝔼 i, f i` is notation for `Finset.expect Finset.univ f`. It is the expectation of `f i` where `i`
+  ranges over the finite domain of `f`.
+* `𝔼 i ∈ s with p i, f i` is notation for `Finset.expect (Finset.filter p s) f`. This is referred to
+  as `expectWith` in lemma names.
+* `𝔼 (i ∈ s) (j ∈ t), f i j` is notation for `Finset.expect (s ×ˢ t) (fun ⟨i, j⟩ ↦ f i j)`.
+-/
+
+open Finset Function
+open Fintype (card)
+open scoped Pointwise
+
+variable {ι κ M N : Type*}
+
+local notation a " /ℚ " q => (q : ℚ≥0)⁻¹ • a
+
+/-- Average of a function over a finset. If the finset is empty, this is equal to zero. -/
+def Finset.expect [AddCommMonoid M] [Module ℚ≥0 M] (s : Finset ι) (f : ι → M) : M :=
+  (s.card : ℚ≥0)⁻¹ • ∑ i ∈ s, f i
+
+namespace BigOperators
+open Batteries.ExtendedBinder Lean Meta
+
+/--
+* `𝔼 i ∈ s, f i` is notation for `Finset.expect s f`. It is the expectation of `f i` where `i`
+  ranges over the finite set `s` (either a `Finset` or a `Set` with a `Fintype` instance).
+* `𝔼 i, f i` is notation for `Finset.expect Finset.univ f`. It is the expectation of `f i` where `i`
+  ranges over the finite domain of `f`.
+* `𝔼 i ∈ s with p i, f i` is notation for `Finset.expect (Finset.filter p s) f`.
+* `𝔼 (i ∈ s) (j ∈ t), f i j` is notation for `Finset.expect (s ×ˢ t) (fun ⟨i, j⟩ ↦ f i j)`.
+
+These support destructuring, for example `𝔼 ⟨i, j⟩ ∈ s ×ˢ t, f i j`.
+
+Notation: `"𝔼" bigOpBinders* ("with" term)? "," term` -/
+scoped syntax (name := bigexpect) "𝔼 " bigOpBinders ("with " term)? ", " term:67 : term
+
+scoped macro_rules (kind := bigexpect)
+  | `(𝔼 $bs:bigOpBinders $[with $p?]?, $v) => do
+    let processed ← processBigOpBinders bs
+    let i ← bigOpBindersPattern processed
+    let s ← bigOpBindersProd processed
+    match p? with
+    | some p => `(Finset.expect (Finset.filter (fun $i ↦ $p) $s) (fun $i ↦ $v))
+    | none => `(Finset.expect $s (fun $i ↦ $v))
+
+open Lean Meta Parser.Term PrettyPrinter.Delaborator SubExpr
+open Batteries.ExtendedBinder
+
+/-- Delaborator for `Finset.expect`. The `pp.piBinderTypes` option controls whether
+to show the domain type when the expect is over `Finset.univ`. -/
+@[scoped delab app.Finset.expect] def delabFinsetexpect : Delab := whenPPOption getPPNotation do
+  let #[_, _, _, s, f] := (← getExpr).getAppArgs | failure
+  guard $ f.isLambda
+  let ppDomain ← getPPOption getPPPiBinderTypes
+  let (i, body) ← withAppArg $ withBindingBodyUnusedName fun i => do
+    return (i, ← delab)
+  if s.isAppOfArity ``Finset.univ 2 then
+    let binder ←
+      if ppDomain then
+        let ty ← withNaryArg 0 delab
+        `(bigOpBinder| $(.mk i):ident : $ty)
+      else
+        `(bigOpBinder| $(.mk i):ident)
+    `(𝔼 $binder:bigOpBinder, $body)
+  else
+    let ss ← withNaryArg 3 $ delab
+    `(𝔼 $(.mk i):ident ∈ $ss, $body)
+
```

Comment:
Can you dedup this with the `Finset.sum` printer somehow?

Reply:
Sounds too fragile. I would need to keep the argument orders in sync (note that `expect` has a `Module _ _` argument which `sum` and `prod` don't have OR make the general delaborator parametric over the argument order, but I don't know how to make a parametric match